### DISCUSSION TITLE
Could bun run in the browser?

### DISCUSSION THREAD

**Discussion by NateRadebaugh** [oven-sh/bun#10400] [Ideas]
I've noticed from performance testing some of my methods using bun vs vitest and wondering, is there any feasible solution for replacing v8 in browser with bun instead? I suppose it'd be a WASM or web worker sort of solution, or would be up to frameworks to support that sort of thing.

Wondering if anyone else has thoughts

**Comment by barak007** [comment]
I would love to see something like a [Web Container  ](https://webcontainers.io/) but for Bun.

**Comment by liudonghua123** [comment]
If bun could be compiled to wasm, then [Web Container/StackBlitz](https://webcontainers.io/), [CodeSandbox](https://github.com/codesandbox) would be easier to implement in Web.

**Comment by guest271314** [comment]
Here's Bun controlled from the browser using Native Messaging https://github.com/guest271314/native-messaging-bun. Bun has a while to go to replace the V8 JavaScript/WebAssembly engine, and Chromium implementation of WHATWG Fetch, WICG Import Maps, etc.

**Comment by guest271314** [comment]
FYI here's Node.js run *in* the browser https://github.com/cemalgnlts/now.

**Comment by theshadowagent** [comment]
Bumping this, this would be awesome!

**Comment by janis-me** [comment]
I would love to get an opinion on this by someone who knows the codebase and the applicability of, for example, compiling bun to webassembly.

Node-js compatible web runtimes like Webcontainers and Nodebox are awesome and with recent investments into tools like bolt.new (They just raised over $100M) it's clear to see that they have usecases. However, both tools are closed-source or at least not usable for production environments, and with bun being the expert in developing Nodejs-compatible APIs, I strongly feel that a first class supported js runtime that executes in the web would be a massive selling point for bun. I would love to contribute on this, but this would probably be a serious effort.

**Reply by theshadowagent** [reply]
Second this

**Reply by nonibytes** [reply]
+1 to this!

**Comment by pukingrainbows** [comment]
Honest question, (by the way I think would be awesome), but what would be the point to compile bun into wasm to run it next to the browser? What would bring `bun` that currently cannot be expose or be consume into a web worker by pure javascript?

For languages like c++, rust, go, c#, zig you are bringing/bridging/exposing their ecosystem with wasm before unreachable for the browsers. 3d engines, audio processing, images processing, etc, etc. 

But bun has nothing specialized part for the browser and if you would need a library you can already bundle and mount it into a web worker. So what would be the gain? what would do bun or bring bun that cannot already be done with pure javascript?

What use cases would we benefiting from requesting this implementation to the bun team? I don't see the ROI to be fair.

> I've noticed from performance testing some of my methods using bun vs vitest and wondering, is there any feasible solution for 
> replacing v8 in browser with bun instead? I suppose it'd be a WASM or web worker sort of solution, or would be up to frameworks to > support that sort of thing.

You could do this by building bun into an executable isn't it, already support it? you don't need wasm or am I missing something?

**Reply by qm3ster** [reply]
https://webcontainers.io/ currently works by compiling nodejs to WASM.
Bun is smaller and faster.

**Comment by SebastienGllmt** [comment]
I created a feature request to add support for WASM components in bun [here](https://github.com/oven-sh/bun/issues/24867). It's tangentially related to this discussion, as WASM components would allow exporting Bun components to WASM in a modular way while still having them be composable (as WASM components are composable using [WAC](https://github.com/bytecodealliance/wac))
